diff -uN ssh2-0.11.3.orig/ssh2_fopen_wrappers.c ssh2-0.11.3/ssh2_fopen_wrappers.c
--- ssh2-0.11.3.orig/ssh2_fopen_wrappers.c	2011-09-27 22:05:05.000000000 +0200
+++ ssh2-0.11.3/ssh2_fopen_wrappers.c	2012-08-10 14:05:23.639357860 +0200
@@ -74,6 +74,31 @@
 	return libssh2_channel_flush_ex(abstract->channel, abstract->streamid);
 }
 
+static int php_ssh2_channel_stream_cast (php_stream *stream, int castas, void** ret TSRMLS_DC) {
+
+       php_ssh2_channel_data *abstract = (php_ssh2_channel_data*)stream->abstract;
+
+       int socket = libssh2_channel_socket_fd(abstract->channel);
+
+       // Load socket
+       switch(castas)  {
+               case PHP_STREAM_AS_FD_FOR_SELECT:
+                       if (ret) {
+                               *(int *)ret = socket;
+                       }
+                       return SUCCESS;
+               case PHP_STREAM_AS_FD:
+               case PHP_STREAM_AS_SOCKETD:
+                       if (ret) {
+                               *(int *)ret = socket;
+                       }
+                       return SUCCESS;
+               default:
+                       return FAILURE;
+       }
+       return SUCCESS;
+}
+
 static int php_ssh2_channel_stream_set_option(php_stream *stream, int option, int value, void *ptrparam TSRMLS_DC)
 {
 	php_ssh2_channel_data *abstract = (php_ssh2_channel_data*)stream->abstract;
@@ -102,7 +127,7 @@
 	php_ssh2_channel_stream_flush,
 	PHP_SSH2_CHANNEL_STREAM_NAME,
 	NULL, /* seek */
-	NULL, /* cast */
+	php_ssh2_channel_stream_cast, /* cast */
 	NULL, /* stat */
 	php_ssh2_channel_stream_set_option,
 };
diff -uN ssh2-0.11.3.orig/ssh2.c ssh2-0.11.3/ssh2.c
--- ssh2-0.11.3.orig/ssh2.c	2011-09-27 22:05:05.000000000 +0200
+++ ssh2-0.11.3/ssh2.c	2012-08-10 14:06:48.770009152 +0200
@@ -427,7 +427,7 @@
 		}
 	}
 
-	if (libssh2_session_startup(session, socket)) {
+	if (libssh2_session_handshake(session, socket)) {
 		int last_error = 0;
 		char *error_msg = NULL;
 
@@ -653,7 +653,7 @@
 	ZEND_FETCH_RESOURCE(session, LIBSSH2_SESSION*, &zsession, -1, PHP_SSH2_SESSION_RES_NAME, le_ssh2_session);
 
 	/* TODO: Support passphrase callback */
-	if (libssh2_userauth_publickey_fromfile_ex(session, username, username_len, pubkey, privkey, passphrase)) {
+	if (libssh2_userauth_publickey_fromfile(session, username, pubkey, privkey, passphrase)) {
 		char *buf;
 		int len;
 		libssh2_session_last_error(session, &buf, &len, 0);
